### PR TITLE
Feature Duplicate Rules

### PR DESCRIPTION
**Pull request by Daniel-Jang**
Original Issue: #296 

Allows the user to duplicate any rule they wish to duplicate. Also, the user can duplicate an or rule group.

Before:
image

After:
image

Or rule group
Before:
image

After:
image

### CODE REVIEW

File: cassdegrees/static/js/staff/rules.js [hunk 1]
```diff
@@ -831,10 +831,20 @@ Vue.component('rule_either_or', {
             this.details.either_or[group].splice(index, 1);
             this.do_redraw();
         },
+        duplicate_rule: function(index, group) {
+            // JSON.parse(JSON.stringify(...)) is done to actually duplicate the contents of the rule, rather than just copying the memory references.
+            this.details.either_or[group].splice(index, 0, JSON.parse(JSON.stringify(this.details.either_or[group][index])));
```
Comment (j-selby) [inline cassdegrees/static/js/staff/rules.js:852]: JSON encoding/decoding is ugly here, but this is JavaScript, so its all ugly anyway :P - does the job anyway.

**Review by j-selby — CHANGES_REQUESTED** [review]
This doesn't appear to work on the courses page - creating a rule and pressing duplicate does nothing, incl. in the console. (Firefox 71.0a1, Windows 10).

**Comment by Daniel-Jang** [comment]
That is interesting, I will fix that, thanks James.

File: cassdegrees/static/js/staff/rules.js (hunk 1 above)
Comment (Daniel-Jang) [reply cassdegrees/static/js/staff/rules.js:852]: HAHA yeah, I did not like this solution but there was no better way...

**Review by JackSekoranja — APPROVED** [review]
Works well, happy for you to merge.

**Review by j-selby — APPROVED** [review]

**Pull request merged**